packages/pds/src/sequencer/outbox.ts

```diff
@@ -0,0 +1,111 @@
+    const { maxBufferSize = 500 } = opts
+    this.cutoverBuffer = []
+    this.outBuffer = new AsyncBuffer<RepoAppendEvent>(maxBufferSize)
+  }
+
+  // event stream occurs in 3 phases
+  // 1. backfill events: events that have been added to the DB since the last time a connection was open.
+  // The outbox is not yet listening for new events from the sequencer
+  // 2. cutover: the outbox has caught up with where the sequencer purports to be,
+  // but the sequencer might already be halfway through sending out a round of updates.
+  // Therefore, we start accepting the sequencer's events in a buffer, while making our own request to the
+  // database to ensure we're caught up. We then dedupe the query & the buffer & stream the events in order
+  // 3. streaming: we're all caught up on historic state, so the sequencer outputs events and we
+  // immediately yield them
+  async *events(backfillFrom?: string): AsyncGenerator<RepoAppendEvent> {
+    // catch up as much as we can
+    if (backfillFrom) {
+      for await (const evt of this.getBackfill(backfillFrom)) {
+        yield evt
+        this.lastSeen = evt.seq
+      }
+    } else {
+      // if not backfill, we don't need to cutover, just start streaming
+      this.caughtUp = true
+    }
+
+    // streams updates from sequencer, but buffers them for cutover as it makes a last request
+    this.sequencer.on('events', (evts) => {
+      if (this.caughtUp) {
+        this.outBuffer.pushMany(evts)
+      } else {
+        this.cutoverBuffer = [...this.cutoverBuffer, ...evts]
+      }
+    })
+
+    const cutover = async () => {
+      // only need to perform cutover if we've been backfilling
+      if (backfillFrom) {
+        const cutoverEvts = await this.sequencer.requestSeqRange({
+          earliestSeq: this.lastSeen,
+        })
+        this.outBuffer.pushMany(cutoverEvts)
+        // dont worry about dupes, we ensure order on yield
+        this.outBuffer.pushMany(this.cutoverBuffer)
+        this.caughtUp = true
+        this.cutoverBuffer = []
+      } else {
+        this.caughtUp = true
+      }
+    }
+    cutover()
+
+    while (true) {
+      try {
+        for await (const evt of this.outBuffer.events()) {
+          if (evt.seq > this.lastSeen) {
+            yield evt
+            this.lastSeen = evt.seq
+          }
+        }
+      } catch (err) {
+        if (err instanceof AsyncBufferFullError) {
+          throw new StreamConsumerTooSlowError(err)
+        } else {
+          throw err
+        }
+      }
+    }
+  }
+
+  // yields only historical events
+  async *getBackfill(startTime?: string) {
+    while (true) {
+      const evts = await this.sequencer.requestSeqRange({
+        earliestTime: startTime,
+        earliestSeq: this.lastSeen,
+        latestSeq: this.sequencer.lastSeen,
+        limit: 50,
+      })
+      for (const evt of evts) {
+        yield evt
+      }
+      if (evts.length < 50) {
+        break
+      }
```

Comment:
I think this break will always occur due to the `limit: 50`.  Not sure I totally track how this works— is it possible it's intended to be `evts.length === 0`?

Reply:
Ohh duh, I get it— if there are less than 50 then you know you got all of them, and it avoids an additional roundtrip to get `0`.  If you get exactly `50` then you keep going.